test/run.py

```diff
@@ -284,3 +284,74 @@ def test_qa_list_of_answers(self):
         self.assertTrue(answer_re.match(output), "'%s' matches pattern '%s'" % (output, answer_re.pattern))
         # restore
         RunQAShort.CYCLE_ANSWERS = orig_cycle_answers
+
+    def test_cmdlist(self):
+        """Tests for CmdList."""
+
+        # starting with empty command is supported
+        # this is mainly useful when parts of a command are put together separately (cfr. mympirun)
+        cmd = CmdList()
+        self.assertEqual(cmd, [])
+        cmd.add('-x')
+        self.assertEqual(cmd, ['-x'])
+
+        cmd = CmdList('test')
+        self.assertEqual(cmd, ['test'])
+
+        # can add options/arguments via string or list of strings
+        cmd.add('-t')
+        cmd.add(['--opt', 'foo', '-o', 'bar', '--optbis=baz'])
+
+        expected = ['test', '-t', '--opt', 'foo', '-o', 'bar', '--optbis=baz']
+        self.assertEqual(cmd, ['test', '-t', '--opt', 'foo', '-o', 'bar', '--optbis=baz'])
+
+        # add options/arguments via a template
+        cmd.add('%(name)s', tmpl_vals={'name': 'namegoeshere'})
+        cmd.add(['%(two)s', '%(one)s'], tmpl_vals={'one': 1, 'two': 2})
+        cmd.add('%(three)s%(five)s%(one)s', tmpl_vals={'one': '1', 'three': '3', 'five': '5'})
+        cmd.add('%s %s %s', tmpl_vals=('foo', 'bar', 'baz'))
+
+        expected.extend(['namegoeshere', '2', '1', '351', 'foo bar baz'])
+        self.assertEqual(cmd, expected)
+
+        # .append and .extend are broken, on purpose, to force use of add
+        self.assertErrorRegex(NotImplementedError, "Use add", cmd.append, 'test')
+        self.assertErrorRegex(NotImplementedError, "Use add", cmd.extend, ['test1', 'test2'])
+
+        # occurence of spaces can be disallowed (but is allowed by default)
+        cmd.add('this has spaces')
+
+        err = "Found one or more spaces"
+        self.assertErrorRegex(ValueError, err, cmd.add, 'this has spaces', allow_spaces=False)
+
+        kwargs = {
+            'tmpl_vals': {'foo': 'this has spaces'},
+            'allow_spaces': False,
+        }
+        self.assertErrorRegex(ValueError, err, cmd.add, '%(foo)s', **kwargs)
+
+        kwargs = {
+            'tmpl_vals': {'one': 'one ', 'two': 'two'},
+            'allow_spaces': False,
+        }
+        self.assertErrorRegex(ValueError, err, cmd.add, '%(one)s%(two)s', **kwargs)
+
+        expected.append('this has spaces')
+        self.assertEqual(cmd, expected)
+
+        # can also init with multiple arguments
+        cmd = CmdList('echo', "hello world", 'test')
+        self.assertEqual(cmd, ['echo', "hello world", 'test'])
+
+        # can also create via template
+        self.assertEqual(CmdList('%(one)s', tmpl_vals={'one': 1}), ['1'])
+
+        # adding non-string items yields an error
+        self.assertErrorRegex(ValueError, "Non-string item", cmd.add, 1)
+        self.assertErrorRegex(ValueError, "Non-string item", cmd.add, None)
+        self.assertErrorRegex(ValueError, "Non-string item", cmd.add, ['foo', None])
```

Comment:
add test for adding list of lists, maybe have it show a clear error like: you can't add a list of lists (something an extend command might be able to do)

Reply:
test case added
a list is not a string, so error is clear imho